File: app/src/main/java/io/intrepid/contest/models/Entry.java
package io.intrepid.contest.models;

import android.util.SparseIntArray;

import java.util.Date;
import java.util.UUID;

public class Entry {
    public UUID id;
    public String title;
    public Date createdAt;
    public Date updatedAt;
    public String photoUrl;
    private transient SparseIntArray mapOfIndexToScore = new SparseIntArray();
    private transient int categoriesSize;

    public Entry() {
        id = UUID.randomUUID();
    }

    public boolean isCompletelyScored() {
        return mapOfIndexToScore.size() == categoriesSize;
    }

    public int getRatingAverage() {
        int sum = 0;
        for (int index = 0; index < mapOfIndexToScore.size(); index++) {
            sum += mapOfIndexToScore.get(index);
        }
        return sum / categoriesSize;
    }

    public void setCategoriesSize(int categoriesSize) {
        this.categoriesSize = categoriesSize;
    }

    public void acceptScore(int position, int newRating) {
        mapOfIndexToScore.put(position, newRating);
    }
}

File: app/src/main/java/io/intrepid/contest/models/Entry.kt
package io.intrepid.contest.models

import android.support.v4.util.SparseArrayCompat
import java.util.*

open class Entry {
    var id: UUID
    var title: String? = null
    var createdAt: Date? = null
    var updatedAt: Date? = null
    var photoUrl: String? = null
    @Transient private val mapOfIndexToScore = SparseArrayCompat<Int>()
    @Transient private var categoriesSize: Int = 0

    init {
        id = UUID.randomUUID()
    }

    val isCompletelyScored: Boolean
        get() = mapOfIndexToScore.size() == categoriesSize

    val ratingAverage: Int
        get() {
            var sum = 0
            for (index in 0..mapOfIndexToScore.size() - 1) {
                sum += mapOfIndexToScore.get(index)
            }
            return sum / categoriesSize
        }

    fun getScoreAt(categoryIndex: Int): Int {
        return mapOfIndexToScore.get(categoryIndex)
    }

    fun setCategoriesSize(categoriesSize: Int) {
        this.categoriesSize = categoriesSize
    }

    fun acceptScore(position: Int, newRating: Int) {
        mapOfIndexToScore.put(position, newRating)
    }
}
